packages/flutter_tools/lib/src/ios/devices.dart

```diff
@@ -74,15 +74,20 @@ class IOSDevices extends PollingDeviceDiscovery {
     // Start by populating all currently attached devices.
     deviceNotifier.updateWithNewList(await pollingGetDevices());
 
-    _observedDeviceEventsSubscription ??= _xcdevice.observedDeviceEvents().listen(
+    // cancel any outstanding subscriptions.
+    await _observedDeviceEventsSubscription?.cancel();
+    _observedDeviceEventsSubscription = _xcdevice.observedDeviceEvents().listen(
       _onDeviceEvent,
       onError: (dynamic error, StackTrace stack) {
         _logger.printTrace('Process exception running xcdevice observe:\n$error');
       }, onDone: () {
-      // If the xcdevice process is killed, restart it.
-      _logger.printTrace('xcdevice observe stopped, restarting');
-      startPolling();
-    },
+        _logger.printTrace('xcdevice observe stopped, restarting');
+
+        // If the xcdevice process is killed, give it a few seconds and restart it.
+        Timer(const Duration(seconds: 1), () {
```

Comment:
What if the device discovery exits immediately on start up (something is broken). Will this keep scheduling work in a loop? Likewise if this is killed as part of tear down.

Reply:
`PollingDeviceDiscovery` polling currently schedules a timer every 4 seconds to call device-specific discovery code, so that's the existing behavior.  It depends what we want the behavior to be.  If I `killall` xcdevice (exits 137), I would expect it to get rescheduled and the IDE not to need to be restarted.  However if it just fails in some other way, rerunning it probably won't help?  @zanderso What do you think?

Restarting after `stopPolling` is bad though...

Reply:
Its a bit of a tricky situation. How common is it for this process to go down during an IDE session? It might be less risky to log a message and move on

Reply:
I never saw it organically go down in my testing, but I also never saw my hard drive fill up with `xcdevice` caches described in #56826.

Good opportunity for metrics?

Reply:
yeah, if we're not sure lets measure

Reply:
It looks like `close()` on the `StreamController` for this `Stream` isn't called, so how does the 'onDone' callback get invoked?

Reply:
It is now.  See comment at https://github.com/flutter/flutter/pull/58137#discussion_r432732711.